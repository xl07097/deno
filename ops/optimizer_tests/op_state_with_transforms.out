#[allow(non_camel_case_types)]
///Auto-generated by `deno_ops`, i.e: `#[op]`
///
///Use `op_now::decl()` to get an op-declaration
///you can include in a `deno_core::Extension`.
pub struct op_now;
#[doc(hidden)]
impl op_now {
    pub const fn name() -> &'static str {
        stringify!(op_now)
    }
    #[allow(clippy::not_unsafe_ptr_arg_deref)]
    pub extern "C" fn v8_fn_ptr<'scope, TP>(
        info: *const deno_core::v8::FunctionCallbackInfo,
    )
    where
        TP: TimersPermission + 'static,
    {
        let info = unsafe { &*info };
        let scope = &mut unsafe { deno_core::v8::CallbackScope::new(info) };
        let args = deno_core::v8::FunctionCallbackArguments::from_function_callback_info(
            info,
        );
        let rv = deno_core::v8::ReturnValue::from_function_callback_info(info);
        Self::v8_func::<TP>(scope, args, rv);
    }
    pub const fn decl<'scope, TP>() -> deno_core::OpDecl
    where
        TP: TimersPermission + 'static,
    {
        deno_core::OpDecl {
            name: Self::name(),
            v8_fn_ptr: Self::v8_fn_ptr::<TP> as _,
            enabled: true,
            fast_fn: {
                use deno_core::v8::fast_api::Type::*;
                use deno_core::v8::fast_api::CType;
                Some(
                    deno_core::v8::fast_api::FastFunction::new(
                        &[V8Value, TypedArray(CType::Uint8), CallbackOptions],
                        CType::Void,
                        op_now_fast_fn::<TP> as *const ::std::ffi::c_void,
                    ),
                )
            },
            is_async: false,
            is_unstable: false,
            is_v8: false,
            force_registration: false,
            arg_count: 1usize as u8,
        }
    }
    #[inline]
    #[allow(clippy::too_many_arguments)]
    pub fn call<TP>(state: &mut OpState, buf: &mut [u8])
    where
        TP: TimersPermission + 'static,
    {}
    pub fn v8_func<'scope, TP>(
        scope: &mut deno_core::v8::HandleScope<'scope>,
        args: deno_core::v8::FunctionCallbackArguments,
        mut rv: deno_core::v8::ReturnValue,
    )
    where
        TP: TimersPermission + 'static,
    {
        let ctx = unsafe {
            &*(deno_core::v8::Local::<deno_core::v8::External>::cast(args.data()).value()
                as *const deno_core::_ops::OpCtx)
        };
        let arg_0 = {
            let value = args.get(0usize as i32);
            match deno_core::v8::Local::<deno_core::v8::ArrayBuffer>::try_from(value) {
                Ok(b) => {
                    let byte_length = b.byte_length();
                    if let Some(data) = b.data() {
                        let store = data.cast::<u8>().as_ptr();
                        unsafe { ::std::slice::from_raw_parts_mut(store, byte_length) }
                    } else {
                        &mut []
                    }
                }
                Err(_) => {
                    if let Ok(view)
                        = deno_core::v8::Local::<
                            deno_core::v8::ArrayBufferView,
                        >::try_from(value) {
                        let len = view.byte_length();
                        let offset = view.byte_offset();
                        let buffer = match view.buffer(scope) {
                            Some(v) => v,
                            None => {
                                return deno_core::_ops::throw_type_error(
                                    scope,
                                    format!("Expected ArrayBufferView at position {}", 0usize),
                                );
                            }
                        };
                        if let Some(data) = buffer.data() {
                            let store = data.cast::<u8>().as_ptr();
                            unsafe {
                                ::std::slice::from_raw_parts_mut(store.add(offset), len)
                            }
                        } else {
                            &mut []
                        }
                    } else {
                        return deno_core::_ops::throw_type_error(
                            scope,
                            format!("Expected ArrayBufferView at position {}", 0usize),
                        );
                    }
                }
            }
        };
        let result = Self::call::<
            TP,
        >(&mut std::cell::RefCell::borrow_mut(&ctx.state), arg_0);
        let op_state = ::std::cell::RefCell::borrow(&*ctx.state);
        op_state.tracker.track_sync(ctx.id);
    }
}
#[allow(clippy::too_many_arguments)]
fn op_now_fast_fn<'scope, TP>(
    _: deno_core::v8::Local<deno_core::v8::Object>,
    buf: *const deno_core::v8::fast_api::FastApiTypedArray<u8>,
    fast_api_callback_options: *mut deno_core::v8::fast_api::FastApiCallbackOptions,
) -> ()
where
    TP: TimersPermission + 'static,
{
    use deno_core::v8;
    use deno_core::_ops;
    let __opts: &mut v8::fast_api::FastApiCallbackOptions = unsafe {
        &mut *fast_api_callback_options
    };
    let __ctx = unsafe {
        &*(v8::Local::<v8::External>::cast(unsafe { __opts.data.data }).value()
            as *const _ops::OpCtx)
    };
    let state = &mut ::std::cell::RefCell::borrow_mut(&__ctx.state);
    let buf = unsafe { (&*buf).get_storage_if_aligned().unwrap_unchecked() };
    let result = op_now::call::<TP>(state, buf);
    result
}
