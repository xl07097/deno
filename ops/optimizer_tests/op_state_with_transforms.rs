pub fn op_now<TP>(state: &mut OpState, buf: &mut [u8])
where
  TP: TimersPermission + 'static,
{
}
