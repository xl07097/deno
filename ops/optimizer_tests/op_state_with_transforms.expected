=== Optimizer Dump ===
returns_result: false
has_ref_opstate: true
has_rc_opstate: false
has_fast_callback_option: false
needs_fast_callback_option: false
fast_result: Some(Void)
fast_parameters: [V8Value, Uint8Array]
transforms: {1: Transform { kind: SliceU8(true), index: 1 }}
is_async: false
fast_compatible: true
